### ISSUE TITLE
[BUG] JSON UI only shows stdin when connecting to an existing session

### ISSUE THREAD

**Issue by QuinnFreedman** [mawww/kakoune#4432]
### Version of Kakoune

v2021.11.08

### Reproducer

Create a script that will print text without exiting:

```sh
# make_commands.sh
while :
do
    echo '{"jsonrpc":"2.0","method":"keys","params":["e"]}'
    sleep 1
done
```

Then, create a named kak session

```sh
kak -ui "json" -s "test"
```

Then, try to connect with a second client that has a non-tty stdin:

```sh
sh ./make_commands.sh | kak -ui "json" -c "test"
```

### Outcome

the second client shows a FIFO buffer where the contents of stdin are displayed as text rather than interpreted as commands.

### Expectations

If you pipe input into a normal kak instance, it will display the text from stdin in a fifo buffer. 
```sh
sh ./make_commands.sh | kak 
```
(The status line shows ` *stdin* 1:1 [scratch][fifo] 1 sel - client0@[1608978]`.)

If you pipe text to stdin of a kak instance created with `-ui json`, it will know to treat the input as JSON RPC commands and open the normal scratch buffer.
```sh
sh ./make_commands.sh | kak -ui "json"
```
```json
{ "jsonrpc": "2.0", "method": "draw_status", "params": [[], [{ "face": { "fg": "default", "bg": "default", "underline": "default", "attributes": [] }, "contents": "*scratch* 1:1 " }, { "face": { "fg": "black", "bg": "yellow", "underline": "default", "attributes": [] }, "contents": "[scratch]" }, { "face": { "fg": "default", "bg": "default", "underline": "default", "attributes": [] }, "contents": " " }, { "face": { "fg": "blue", "bg": "default", "underline": "default", "attributes": [] }, "contents": "1 sel" }, { "face": { "fg": "default", "bg": "default", "underline": "default", "attributes": [] }, "contents": " - client0@[1610796]" }], { "fg": "cyan", "bg": "default", "underline": "default", "attributes": [] }] }
```

But, if you pipe into the stdin of a JSON UI client that is connecting to an existing session, it goes back to the original behaviour where the input is interpreted as text, not commands, and is displayed (the status line is `[fifo]`).

```sh
sh ./make_commands.sh | kak -ui "json" -c "test"
```
```json
{ "jsonrpc": "2.0", "method": "draw_status", "params": [[], [{ "face": { "fg": "default", "bg": "default", "underline": "default", "attributes": [] }, "contents": "*stdin-1* 1:1 " }, { "face": { "fg": "black", "bg": "yellow", "underline": "default", "attributes": [] }, "contents": "[scratch][fifo]" }, { "face": { "fg": "default", "bg": "default", "underline": "default", "attributes": [] }, "contents": " " }, { "face": { "fg": "blue", "bg": "default", "underline": "default", "attributes": [] }, "contents": "1 sel" }, { "face": { "fg": "default", "bg": "default", "underline": "default", "attributes": [] }, "contents": " - client1@[test]" }], { "fg": "cyan", "bg": "default", "underline": "default", "attributes": [] }] }
```

This makes multiple clients useless for JSON UI because you cannot interact with them by sending commands. Instead, it should interpret stdin as RPC commands and not display them to the user

### Additional information

I am using Ubuntu 20.04.2 LTS.

I first encountered this error running kak from a Rust program with 

```rust
Command::new("kak")
    .args(&["-ui", "json"])
    .args(&["-c", "test"])
    .stdout(Stdio::piped())
    .stdin(Stdio::piped())
    .spawn()
    .unwrap()
```
but I was able to reproduce it in the terminal (Alacritty) as well.

This error seems to be a regression. A very old version of kak (probably from a couple years ago -- I don't remember the version number) did not have this issue but a late 2020 or early 2021 version introduced it.

**Comment by sidkshatriya** [comment]
> This makes multiple clients useless for JSON UI because you cannot interact with them by sending commands. 

The following works:

```bash
$ echo 'eval -client client0 -verbatim -- execute-keys %{itest from the command line}' | kak -p test
```

You should see the JSON UI update and show you that it has the text "test from the command line" in the scratch buffer.

Here you are sending the kakoune client `client0` within the `test` kakoune session some commands. This is exactly how you would do this without a json-rpc interface. 

I'm not really familiar with the JSON UI approach but I take the guess here is that there is only _one_ way of interacting with the client UI itself. When you start up a session of kakoune normally there is only one thing interacting with the UI i.e. your terminal.

Everything else is sending commands to the server itself via `kak -p` and this is no different for the JSON UI.

---

However, there _is_ a workaround for doing things the way you want.
 
```bash
$ mkfifo /tmp/testsessionrpc
$ kak -ui json -s test </tmp/testsessionrpc
```

Now you can have multiple programs interacting with this UI.

In one console window do
```
$ sh ./make_commands.sh >/tmp/testsessionrpc
```

In another console window you can do another
```
$ sh ./make_commands.sh >/tmp/testsessionrpc
```

You can ctrl+C when you want. You should notice that the kakoune ui only closes when the last "client" closes the named fifo.

**Caveats**
Here there are 2 processes simultaneously accessing the fifo. I see two problems:

(1) They need to coordinate with each other e.g. one program may send commands to exit the insert mode while the other program may still think that kakoune client is in insert mode
(2) Writes to fifo are NOT atomic and are interleaved if you are writing more than  `PIPE_BUF` bytes. In other words, if your json rpc calls are "large" then they could be arbitrarily be interleaved with other JSON rpc calls. If your json rpc calls are small then you will NOT have this problem.

> Write requests of {PIPE_BUF} bytes or less shall not be interleaved with data from other processes doing writes on the same pipe. Writes of greater than {PIPE_BUF} bytes may have data interleaved, on arbitrary boundaries, with writes by other processes, whether or not the O_NONBLOCK flag of the file status flags is set.

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/write.html
and https://man7.org/linux/man-pages/man7/pipe.7.html

(from the second link PIPE_BUF for linux is 4096 bytes)

_If you expect each jsonrpc call to end before the other starts then you will not face the above two problems_

**Comment by QuinnFreedman** [comment]
@sidkshatriya thanks for the info. 

My usecase is trying to make a GUI frontend for kak that can handle multiple windows/splits in one frame. I wanted to use a different kak client for each split and send the user's keystrokes to the corresponding client depending on which window they have selected. (i.e. the behaviour of `:new` in the terminal)

`kak -p` does allow me to do what I wanted to do, but it seems very weird to me that there would be the whole JSON RPC API for the main client but need to use a different syntax to send keystrokes to other clients. Also, I have some performance concerns about starting a `kak -p` process for every keystroke, but maybe that's unsubstantiated.

I think the FIFO solution does not do what I need to do. My problem is trying to control multiple kak clients from one process, rather than controlling one kak client from multiple processes.

I guess I'm just trying to figure out if this is a bug or intended behaviour, and if it intended, why? If the JSON RPC can't be used for what I'm trying to use it for, what would it be used for?

**Comment by QuinnFreedman** [comment]
I wonder if there is a way to specify which client you want to target when sending JSON RPC commands to the main process. That would work just as well as sending to commands to the stdin of the corresponding client, but if so it does not seem to be documented anywhere in the [JSON UI docs](https://github.com/mawww/kakoune/blob/master/doc/json_ui.asciidoc).

**Comment by sidkshatriya** [comment]
> My problem is trying to control multiple kak clients from one process

To interact with multiple clients in the _same_ session, why not do the following:

```bash
$ kak -ui json -s test
```

Now you can connect **multiple** json ui clients to the **same** session

Do this in different console windows:
```bash
$ kak -ui json -c test
```
```bash
$ kak -ui json -c test
```
```bash
$ kak -ui json -c test
```

Each of them is a different client attached to the _same_ session. Each of them has a stdin with which you can interact with. Each of them has UI which you can interact with over JSON-UI RPC.

Think JSON-UI as interacting with something in headless mode. So you only send things like keys, mouse etc. "virtually"

**Comment by sidkshatriya** [comment]
In other words, for the GUI frontend you are making, you will be spawning multiple kak processes from your "GUI program". You will have the stdin/stdout for each kakoune client (and all of them will be part of the same session) and you can interact with them.

You _wouldnt_ need to spawn a whole kak process to send a single key stroke in this way.

**Comment by Screwtapello** [comment]
> Now you can connect multiple json ui clients to the same session.

...yes, but nobody wants to type JSON blobs manually, you want to generate them in some other program and pipe them to the JSON UI client, which is when you hit this bug.

Looking into the code, it seems that when started normally, `main()` calls `run_server()` which calls `create_local_ui()` which does the right thing - it handles reading stdin into a buffer for the terminal UI, but otherwise does nothing special.

Meanwhile, when started with `-c`, `main()` calls `run_client()` which calls `make_ui()` (no special treatment for any kind of UI) and then `RemoteClient()` which passes the stdin file descriptor to the server to read it.

This is definitely a problem, but it's not obvious to me how to solve it.

**Comment by sidkshatriya** [comment]
@Screwtapello Maybe I'm not able to adequately understand the problem. I feel this should be possible.

Let say you are in this GUI that @QuinnFreedman is building. You want people to be able to type in `:new` in this GUI and get another window that appears along side?

So in that case you simply override the `new` command in your custom configuration. In this (custom) `new` command:
- Your GUI program starts up the equivalent of `kak -ui json -c <your-session`
- Your GUI program has access to the stdin and stdout of the newly started client
- Your GUI can now present a new "window" representing this client

**Comment by QuinnFreedman** [comment]
@sidkshatriya That is exactly what I am doing. Like you said, I override the `new` command to call back to my Rust app (via a custom fifo queue for now, although it would be nice if there was support for arbitrary UI commands in the JSON API). In my Rust app, I start a process with `kak -ui json -c <session name>` and pipe JSON RPC commands to it. The bug seems to be that instances started with `kak -ui json -c` don't correctly handle JSON input.

I get valid JSON UI output back that I can display, but when I send JSON commands via the stdin of the `kak -ui json -c` they are not interpreted correctly.

**Comment by QuinnFreedman** [comment]
@Screwtapello without having looked at the code myself, it sounds like a refactor is needed to move the logic for handling stdin into a function that can be called by both paths. Alternatively, as a stopgap measure, we could add a command line flag to manually specify how to handle stdin rather than trying to detect it based on the UI type

**Comment by sidkshatriya** [comment]
@QuinnFreedman

This was a very intriguing experiment you have posted and I came back to this again. Before diving into the code to try to undertstand its perceived issues, I thought I'd run the experiment in a detailed fashion. I am getting some interesting (and positive) results that seem to be at variance with your findings.

Again, I apologize if I am missing something. 

### Step 1 
First, I modified your script a bit: I changed "e" -> "a". Basically after the first "a" key press (which means append and will cause kakoune to enter _insert_ mode), the rest of the times the script will just add the character `a`. 

_One other advantage of this is that it is quite visual to see the repeated `a`'s in the json-rpc. After a few `a`'s they are easy to identify._ 

```bash
# infinite_append.sh
while :
do
    echo '{"jsonrpc":"2.0","method":"keys","params":["a"]}'
    sleep 1
done
```

### Step 2  _(Console Window 1)_

In a new console window 

```bash
$ tty
/dev/pts/<some-number>
$ kak -ui json -s testing12
```
(this is client0@testing12)

### Step 3  _(Console Window 2)_
In a new console window

```bash
$ tty
# Sanity check, you should have a different number from step 2
/dev/pts/<some-number>
$ sh ./infinite_append | kak -ui json -c testing12
```
(this is client1@testing12)

**Result**
You will see that both windows show an ever increasing stream of content of `a`'s.

**Conclusion:** The json-rpc instructions to the second client are also being interpreted correctly. 

_Notes_
- I never saw `fifo` in my status line during testing
- I was able to repeat this in gnome terminal _and_ tmux when running within gnome terminal

**Comment by QuinnFreedman** [comment]
Hi @sidkshatriya 

Thanks for looking into this. That's very weird that that issue didn't reproduce for you. That sounds like exactly what I was doing. I was originally getting the issue in i3 + Alacritty on Ubuntu 20.04 with kak version `v2021.11.08`. I just copied and pasted your exact steps and still got the same error in Alacritty and Gnome Terminal in native Ubuntu, and in tmux on WSL with kak version `v2021.08.28-93-g6e635ea3`.

For me, the output in the first terminal doesn't change, while the second terminal continues to output a bunch of json strings as text.

When I run `sh infinite_append.sh | kak -ui json -c testing12` in the second terminal, It first sends a status line like this:

```json
{
  "jsonrpc": "2.0",
  "method": "draw_status",
  "params": [
    [],
    [
      {
        "face": {
          "fg": "default",
          "bg": "default",
          "underline": "default",
          "attributes": []
        },
        "contents": "*stdin-0* 1:1 " // <- displaying stdin
      },
      {
        "face": {
          "fg": "black",
          "bg": "yellow",
          "underline": "default",
          "attributes": []
        },
        "contents": "[scratch][fifo]" // <- fifo buffer
      },
      {
        "face": {
          "fg": "default",
          "bg": "default",
          "underline": "default",
          "attributes": []
        },
        "contents": " "
      },
      {
        "face": {
          "fg": "blue",
          "bg": "default",
          "underline": "default",
          "attributes": []
        },
        "contents": "1 sel"
      },
      {
        "face": {
          "fg": "default",
          "bg": "default",
          "underline": "default",
          "attributes": []
        },
        "contents": " - client1@[testing12]"
      }
    ],
    {
      "fg": "cyan",
      "bg": "default",
      "underline": "default",
      "attributes": []
    }
  ]
}
```

Then, after letting it run for a few seconds, I get something like this:

```json
{
  "jsonrpc": "2.0",
  "method": "draw",
  "params": [
    [
      [
        {
          "face": {
            "fg": "black",
            "bg": "white",
            "underline": "default",
            "attributes": [
              "final_fg",
              "final_bg"
            ]
          },
          "contents": "{"
        },
        {
          "face": {
            "fg": "default",
            "bg": "default",
            "underline": "default",
            "attributes": []
          },
          "contents": "\"jsonrpc\":\"2.0\",\"method\":\"keys\",\"params\":[\"a\"]}\\u000a" // <- Showing JSON RPC command as if it were text in the buffer
        }
      ],
      [
        {
          "face": {
            "fg": "default",
            "bg": "default",
            "underline": "default",
            "attributes": []
          },
          "contents": "{\"jsonrpc\":\"2.0\",\"method\":\"keys\",\"params\":[\"a\"]}\\u000a"
        }
      ],
      [
        {
          "face": {
            "fg": "default",
            "bg": "default",
            "underline": "default",
            "attributes": []
          },
          "contents": "{\"jsonrpc\":\"2.0\",\"method\":\"keys\",\"params\":[\"a\"]}\\u000a"
        }
      ],
      [
        {
          "face": {
            "fg": "default",
            "bg": "default",
            "underline": "default",
            "attributes": []
          },
          "contents": "{\"jsonrpc\":\"2.0\",\"method\":\"keys\",\"params\":[\"a\"]}\\u000a"
        }
      ],
      [
        {
          "face": {
            "fg": "default",
            "bg": "default",
            "underline": "default",
            "attributes": []
          },
          "contents": "{\"jsonrpc\":\"2.0\",\"method\":\"keys\",\"params\":[\"a\"]}\\u000a"
        }
      ],
      [
        {
          "face": {
            "fg": "default",
            "bg": "default",
            "underline": "default",
            "attributes": []
          },
          "contents": "\\u000a"
        }
      ]
    ],
    {
      "fg": "default",
      "bg": "default",
      "underline": "default",
      "attributes": []
    },
    {
      "fg": "blue",
      "bg": "default",
      "underline": "default",
      "attributes": []
    }
  ]
}
```

I'm very confused as to what else to do to try to reproduce this, other than to make sure you have a relatively new version of kakoune since I know this is a new-ish bug.

**Comment by sidkshatriya** [comment]
Interesting and puzzling. I'll investigate further to see how I can cause it to _not_ to run properly. 

BTW I'm on e7100dc87434d933fd8268e5bf70080b627750c5 which is one commit after `v2021.11.08`

**Comment by sidkshatriya** [comment]
To make sure that nothing you are doing is getting messed up by your `kakrc` in Step 2 please add `-n` to your command line i.e.

```bash
$ kak -n -ui json -s testing12
```

The experiment should still work with `-n`

**Comment by sidkshatriya** [comment]
Oops. This is my bad -- I forgot to mention that I was running kak with slightly modified `run_client()` code. 

**After you make these modifications this should work for your too** -- please let me know.

```diff
diff --git a/src/main.cc b/src/main.cc
index 6da6e648..a131755a 100644
--- a/src/main.cc
+++ b/src/main.cc
@@ -692,14 +692,14 @@ int run_client(StringView session, StringView name, StringView client_init,
     try
     {
         Optional<int> stdin_fd;
-        if (not isatty(0))
+        /*if (not isatty(0))
         {
             // move stdin to another fd, and restore tty as stdin
             stdin_fd = dup(0);
             int tty = open("/dev/tty", O_RDONLY);
             dup2(tty, 0);
             close(tty);
-        }
+        }*/
 
         EventManager event_manager;
         RemoteClient client{session, name, make_ui(ui_type), getpid(), get_env_vars(),
```

**A more permanent fix will need to be found but this quick fix is what allows my experiment to work**

**Comment by QuinnFreedman** [comment]
Yes, it looks like commenting out that if block fixes the issue! As a medium-term solution just having a cli flag to skip that check would solve my problem. But ideally that behaviour should be fixed. For now I can just comment it out.
Thanks!

**Comment by sidkshatriya** [comment]
The most appropriate fix to be implemented needs to be figured. I've proposed a fix in #4438

However, yes, for your use case commenting out the `if` statement as show [above](https://github.com/mawww/kakoune/issues/4432#issuecomment-968247010) does the trick well.

---
**Comments about this bug**

When a Kakoune client is connecting to an existing session you may / or may not be piping in data.

(a) If nothing is being pipe-d in, the stdin _is_ a tty. The `if (not isatty(0))` check fails and everthing is OK.

(b) However, if something is being piped in that means that the stdin of kakoune is now _not_ a tty. Kakoune assumes that the interesting inputs will actually be coming in from the terminal. So what it does is that it makes the pseudo terminal from which kakoune is started the "real" stdin (actually ignoring the user's explicit intentions). It sends the original stdin (the file descriptor of the read end of the pipe) to the server via a socket (it's possible to transfer file descriptors via socket in unix) where it is proceeded to take input into the fifo.

https://github.com/mawww/kakoune/blob/47c0d2038807757207cde3b3e00fbc1ca9448178/src/remote.cc#L792-L804

This fd arrives via socket recvmsg via a MessageType::Connect message in the above piece of code. If a fd has arrived you will notice that kakoune calls `create_fifo_buffer` -- this is the problem you were running into.

**Comments related to json ui mode**

One interesting thing you will notice in the case where we have not fiddled with kakoune code (i.e. nothing is patched) that the json ui will accept proper json rpc commands. But that is useless to us because no one is actually typing in commands in JSON. That does not happen. But there is a workaround...

**Workaround 1**
If you absolutely did not want to accept any patching of kakoune, you could open `kak -c <session> -ui json` in a `forkpty()`. You will get a fd to which you can read and write json-rpc.

**Workaround 2**
This is the workaround that the above patch employs.  The `if` check is commented out. We ignore the tty so we  _dont_ make it stdin. No auxillary fd is sent via socket. The pipe accepts the json rpc commands instead of the tty.

One small disadvantage (though hardly a big one) with this workaround is that you wont be able to manually type in json rpc commands in the terminal. This is sometimes useful though for debugging. However, since you're in control of the pipe you can send your test commands from there. 

Anyways, since you're automating this in Rust you're probably not interested in this feature anyways...

**Proposed Solution**
See #4438
